docs/proposals/connector-protocol-v2.md

```diff
@@ -224,58 +226,53 @@ notify(StreamId, StreamName, PointOfRef) ->
         point_of_reference(PointOfRef)
     ]).
 
--spec message(Message, MessageId, StreamId, Flags) -> frame() when
-    Message :: boundary | {non_neg_integer(), iodata()} | iodata(),
-    MessageId :: non_neg_integer(),
+-spec message(Flags, StreamId, MessageId, EventTime, Key, Message) -> frame()
+  when
+    Flags :: non_neg_integer(),      % See message bit flag definitions above
     StreamId :: non_neg_integer(),
-    Flags :: non_neg_integer(). % See message bit flag definitions above
-message(_Message=boundary, MessageId, StreamId, Flags) ->
-    frame([
-        ?MESSAGE,
-        u64(StreamId),
-        u16(Flags bor ?BOUNDARY),
-        u64(MessageId)
-    ]);
-message({EventTime, Message}, MessageId, StreamId, Flags) ->
-    frame([
-        ?MESSAGE,
-        u64(StreamId),
-        u16(Flags bor ?EVENT_TIME),
-        u64(MessageId),
-        u64(EventTime),
-        Message
-    ]);
-message(Message, MessageId, StreamId, Flags) ->
+    MessageId :: non_neg_integer(),
+    EventTime :: non_neg_integer(),
+    Key :: string(),
+    Message :: iodata().
+message(Flags, StreamId, MessageId, EventTime, Key, Message) ->
     frame([
-        ?MESSAGE,
-        u64(StreamId),
         u16(Flags),
-        u64(MessageId),
-        Message
+        u64(StreamId),
+        if Flags band ?EPHEMERAL /= 0 then
+            <<>>
+        else
+            u64(MessageId)
+        end,
+        if Flags band ?EVENT_TIME /= 0 then
+            u64(EventTime)
+        else
+            <<>>
+        end,
+        if Flags band ?KEY /= 0 then
+            short_bytes(Key)
+        else
+            <<>>
+        end,
+        if Flags band ?BOUNDARY /= 0 then
+            <<>>
+        else
+            Message
+        end
     ]).
 
 ack(Credits, MessageAcks) ->
     framed([
         ?ACK,
         u32(Credits),
-        lists:map(fun ({StreamId, MessageId}) ->
-            [u64(StreamId), u64(MessageId)]
+        u32(length(MessageAcks)),
+        lists:map(fun ({StreamId, PointofRef}) ->
+            [u64(StreamId), u64(PointOfRef)]
         end, MessageAcks)
     ]).
 
-nack(Credits, StreamId, undefined) ->
-    frame([
-        ?NACK,
-        u32(Credits),
-        u64(StreamId),
-        point_of_reference(0)
-    ]);
-nack(Credits, StreamId, PointOfRef) ->
+nack() ->
     frame([
-        ?NACK,
-        u32(Credits),
-        u64(StreamId),
-        point_of_reference(PointOfRef)
+        ?NACK
```

Comment:
If we terminate on nack, wouldn't closing the TCP socket suffice? Or otherwise reuse the error frame so we can pass a reason along?

Reply:
The reason for sending the NACK is when the cluster's internal state is changing, and this likely-too-simple async protocol can't cope without a restart from the beginning.  The event will be rare but not never.